File: aac/api/python/wia_aac/adapters/base_adapter.py
"""
WIA AAC Base Adapter
Abstract base class for all sensor adapters
"""

import time
from abc import ABC, abstractmethod
from typing import Optional, List, Callable, Any, Dict

from ..types import (
    WiaAacSignal,
    DeviceInfo,
    SensorType,
    SensorConfig,
    SensorOptions,
    SensorOptionDescriptor,
    Timestamp,
    SignalMeta
)

# Signal handler type
SignalHandler = Callable[[WiaAacSignal], None]


class BaseSensorAdapter(ABC):
    """Abstract base class for sensor adapters"""

    def __init__(self):
        self._device_info: Optional[DeviceInfo] = None
        self._is_connected = False
        self._last_signal: Optional[WiaAacSignal] = None
        self._signal_handlers: List[SignalHandler] = []
        self._config: Optional[SensorConfig] = None
        self._options: Dict[str, Any] = {}

    @property
    @abstractmethod
    def type(self) -> str:
        """Sensor type"""
        pass

    @property
    def device_info(self) -> Optional[DeviceInfo]:
        """Device information"""
        return self._device_info

    @property
    def is_connected(self) -> bool:
        """Connection status"""
        return self._is_connected

    async def connect(self, config: SensorConfig) -> None:
        """Connect to sensor"""
        self._config = config
        self._device_info = DeviceInfo(
            manufacturer=config.device.manufacturer if config.device else "Unknown",
            model=config.device.model if config.device else self.type,
            firmware="1.0.0",
            serial=config.device.serial if config.device else None
        )
        self._is_connected = True

    async def disconnect(self) -> None:
        """Disconnect from sensor"""
        self._is_connected = False
        self._device_info = None
        self._config = None
        self._last_signal = None

    def get_last_signal(self) -> Optional[WiaAacSignal]:
        """Get last signal"""
        return self._last_signal

    def on_signal(self, handler: SignalHandler) -> None:
        """Register signal handler"""
        self._signal_handlers.append(handler)

    def off_signal(self, handler: SignalHandler) -> None:
        """Unregister signal handler"""
        if handler in self._signal_handlers:
            self._signal_handlers.remove(handler)

    def configure(self, options: SensorOptions) -> None:
        """Configure sensor options"""
        if hasattr(options, '__dict__'):
            self._options.update({k: v for k, v in options.__dict__.items() if v is not None})
        elif isinstance(options, dict):
            self._options.update(options)

    @abstractmethod
    def get_supported_options(self) -> List[SensorOptionDescriptor]:
        """Get supported options"""
        pass

    def _emit_signal(self, signal: WiaAacSignal) -> None:
        """Emit signal to handlers"""
        self._last_signal = signal
        for handler in self._signal_handlers:
            try:
                handler(signal)
            except Exception as e:
                print(f"Error in signal handler: {e}")

    def _create_signal(
        self,
        sensor_type: str,
        data: Dict[str, Any],
        confidence: float = 0.9
    ) -> WiaAacSignal:
        """Create a signal object"""
        now_ms = int(time.time() * 1000)
        return WiaAacSignal(
            version="1.0.0",
            type=SensorType(sensor_type) if sensor_type in [e.value for e in SensorType] else SensorType.CUSTOM,
            timestamp=Timestamp(
                unix_ms=now_ms,
                iso8601=time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime(now_ms / 1000))
            ),
            device=self._device_info or DeviceInfo(manufacturer="Unknown", model=sensor_type),
            data=data,
            sequence=now_ms,
            meta=SignalMeta(confidence=confidence, validity=True)
        )

File: aac/api/typescript/src/adapters/BaseAdapter.ts
/**
 * WIA AAC Base Adapter
 * Abstract base class for all sensor adapters
 */

import {
  WiaAacSignal,
  DeviceInfo,
  SensorType,
  SensorConfig,
  SensorOptions,
  SensorOptionDescriptor
} from '../types';

export type SignalHandler = (signal: WiaAacSignal) => void;

export interface ISensorAdapter {
  readonly type: SensorType;
  readonly deviceInfo: DeviceInfo | null;
  readonly isConnected: boolean;

  connect(config: SensorConfig): Promise<void>;
  disconnect(): Promise<void>;

  getLastSignal(): WiaAacSignal | null;
  onSignal(handler: SignalHandler): void;
  offSignal(handler: SignalHandler): void;

  configure(options: SensorOptions): void;
  getSupportedOptions(): SensorOptionDescriptor[];
}

export abstract class BaseAdapter implements ISensorAdapter {
  abstract readonly type: SensorType;

  protected _deviceInfo: DeviceInfo | null = null;
  protected _isConnected = false;
  protected _lastSignal: WiaAacSignal | null = null;
  protected signalHandlers: Set<SignalHandler> = new Set();
  protected config: SensorConfig | null = null;
  protected options: SensorOptions = {};

  get deviceInfo(): DeviceInfo | null {
    return this._deviceInfo;
  }

  get isConnected(): boolean {
    return this._isConnected;
  }

  async connect(config: SensorConfig): Promise<void> {
    this.config = config;
    this._deviceInfo = {
      manufacturer: config.device?.manufacturer || 'Unknown',
      model: config.device?.model || this.type,
      firmware: '1.0.0',
      serial: config.device?.serial
    };
    this._isConnected = true;
  }

  async disconnect(): Promise<void> {
    this._isConnected = false;
    this._deviceInfo = null;
    this.config = null;
    this._lastSignal = null;
  }

  getLastSignal(): WiaAacSignal | null {
    return this._lastSignal;
  }

  onSignal(handler: SignalHandler): void {
    this.signalHandlers.add(handler);
  }

  offSignal(handler: SignalHandler): void {
    this.signalHandlers.delete(handler);
  }

  configure(options: SensorOptions): void {
    this.options = { ...this.options, ...options };
  }

  abstract getSupportedOptions(): SensorOptionDescriptor[];

  protected emitSignal(signal: WiaAacSignal): void {
    this._lastSignal = signal;
    for (const handler of this.signalHandlers) {
      try {
        handler(signal);
      } catch (error) {
        console.error('Error in signal handler:', error);
      }
    }
  }

  protected createSignal<T extends SensorType>(
    type: T,
    data: WiaAacSignal['data'],
    confidence = 0.9
  ): WiaAacSignal {
    return {
      version: '1.0.0',
      type,
      timestamp: {
        unix_ms: Date.now(),
        iso8601: new Date().toISOString()
      },
      sequence: Date.now(),
      device: this._deviceInfo || { manufacturer: 'Unknown', model: type },
      data,
      meta: {
        confidence,
        validity: true
      }
    } as WiaAacSignal;
  }
}
